File: algorithmic_toolbox/week_3/02_greedy_algorithms_starter_files/different_summands/different_summands.cpp
#include <iostream>
#include <vector>

using std::vector;

vector<int> optimal_summands(int n) {
    vector<int> summands;
    for (int i = 1; i <= n; i++) {
        n -= i;
        if (n <= i) {
            summands.push_back(n + i);
        } else if (n == 0) {
            summands.push_back(i);
            break;
        } else {
            summands.push_back(i);
        }
    }
    return summands;
}

int main() {
    int n;
    std::cin >> n;
    vector<int> summands = optimal_summands(n);
    std::cout << summands.size() << '\n';
    for (size_t i = 0; i < summands.size(); ++i) {
        std::cout << summands[i] << ' ';
    }
}

File: algorithmic_toolbox/week_3/02_greedy_algorithms_starter_files/different_summands/different_summands.py
# Uses python3
import sys


def optimal_summands(n):
    summands = []
    for i in range(1, n + 1):
        n -= i
        if n <= i:
            summands.append(n + i)
            break
        elif n == 0:
            summands.append(i)
            break
        else:
            summands.append(i)
    return summands


if __name__ == '__main__':
    input = sys.stdin.read()
    n = int(input)
    summands = optimal_summands(n)
    print(len(summands))
    for x in summands:
        print(x, end=' ')
